src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Patterns.cs

```diff
@@ -1,164 +1,581 @@
+                    // PROTOTYPE(patterns2): could handle constant expressions more efficiently.
+                    if (self._sideEffectBuilder.Count != 0)
+                    {
+                        expression = self._factory.Sequence(ImmutableArray<LocalSymbol>.Empty, self._sideEffectBuilder.ToImmutable(), expression);
+                        self._sideEffectBuilder.Clear();
+                    }
+
+                    self._conjunctBuilder.Add(expression);
                 }
 
-                Debug.Assert((object)loweredPattern.Variable != null && loweredInput.Type.Equals(loweredPattern.Variable.GetTypeOrReturnType(), TypeCompareKind.AllIgnoreOptions));
+                var input = _tempAllocator.GetTemp(decision.Input);
+                switch (decision)
+                {
+                    case BoundNonNullDecision d:
+                        // If the actual input is a constant, short-circuit this test
+                        if (d.Input == _inputTemp && _loweredInput.ConstantValue != null)
+                        {
+                            var decisionResult = _loweredInput.ConstantValue != ConstantValue.Null;
+                            if (!decisionResult)
+                            {
+                                _conjunctBuilder.Add(_factory.Literal(decisionResult));
+                                return true;
+                            }
+                        }
+                        else
+                        {
+                            // PROTOTYPE(patterns2): combine null test and type test when possible for improved code
+                            addConjunct(ref this, _localRewriter.MakeNullCheck(d.Syntax, input, input.Type.IsNullableType() ? BinaryOperatorKind.NullableNullNotEqual : BinaryOperatorKind.NotEqual));
+                        }
+                        break;
+                    case BoundTypeDecision d:
+                        {
+                            addConjunct(ref this, _factory.Is(input, d.Type));
+                        }
+                        break;
+                    case BoundValueDecision d:
+                        // If the actual input is a constant, short-circuit this test
+                        if (d.Input == _inputTemp && _loweredInput.ConstantValue != null)
+                        {
+                            var decisionResult = _loweredInput.ConstantValue == d.Value;
+                            if (!decisionResult)
+                            {
+                                _conjunctBuilder.Add(_factory.Literal(decisionResult));
+                                return true;
+                            }
+                        }
+                        else if (d.Value == ConstantValue.Null)
+                        {
+                            addConjunct(ref this, _localRewriter.MakeNullCheck(d.Syntax, input, input.Type.IsNullableType() ? BinaryOperatorKind.NullableNullEqual : BinaryOperatorKind.Equal));
```

Comment:
>input.Type.IsNullableType() [](start = 96, length = 27)

Is this condition ever true?

Reply:
Yes, the input to an is-expression is permitted to be a nullable type.

---
In reply to: [156558837](https://github.com/dotnet/roslyn/pull/23209#discussion_r156558837) [](ancestors = 156558837)